<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkML0101ENSkillsNetwork20718538-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>

<h1 align="center"><font size="5">Final Project: Classification with Python</font></h1>


<h2>Table of Contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
    <li><a href="https://#Section_1">Instructions</a></li>
    <li><a href="https://#Section_2">About the Data</a></li>
    <li><a href="https://#Section_3">Importing Data </a></li>
    <li><a href="https://#Section_4">Data Preprocessing</a> </li>
    <li><a href="https://#Section_5">One Hot Encoding </a></li>
    <li><a href="https://#Section_6">Train and Test Data Split </a></li>
    <li><a href="https://#Section_7">Train Logistic Regression, KNN, Decision Tree, SVM, and Linear Regression models and return their appropriate accuracy scores</a></li>
</a></li>
</div>
<p>Estimated Time Needed: <strong>180 min</strong></p>
</div>

<hr>


# Instructions


In this notebook, you will  practice all the classification algorithms that we have learned in this course.


Below, is where we are going to use the classification algorithms to create a model based on our training data and evaluate our testing data using evaluation metrics learned in the course.

We will use some of the algorithms taught in the course, specifically:

1. Linear Regression
2. KNN
3. Decision Trees
4. Logistic Regression
5. SVM

We will evaluate our models using:

1.  Accuracy Score
2.  Jaccard Index
3.  F1-Score
4.  LogLoss
5.  Mean Absolute Error
6.  Mean Squared Error
7.  R2-Score

Finally, you will use your models to generate the report at the end. 


# About The Dataset


The original source of the data is Australian Government's Bureau of Meteorology and the latest data can be gathered from [http://www.bom.gov.au/climate/dwo/](http://www.bom.gov.au/climate/dwo/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkML0101ENSkillsNetwork20718538-2022-01-01).

The dataset to be used has extra columns like 'RainToday' and our target is 'RainTomorrow', which was gathered from the Rattle at [https://bitbucket.org/kayontoga/rattle/src/master/data/weatherAUS.RData](https://bitbucket.org/kayontoga/rattle/src/master/data/weatherAUS.RData?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkML0101ENSkillsNetwork20718538-2022-01-01)




This dataset contains observations of weather metrics for each day from 2008 to 2017. The **weatherAUS.csv** dataset includes the following fields:

| Field         | Description                                           | Unit            | Type   |
| ------------- | ----------------------------------------------------- | --------------- | ------ |
| Date          | Date of the Observation in YYYY-MM-DD                 | Date            | object |
| Location      | Location of the Observation                           | Location        | object |
| MinTemp       | Minimum temperature                                   | Celsius         | float  |
| MaxTemp       | Maximum temperature                                   | Celsius         | float  |
| Rainfall      | Amount of rainfall                                    | Millimeters     | float  |
| Evaporation   | Amount of evaporation                                 | Millimeters     | float  |
| Sunshine      | Amount of bright sunshine                             | hours           | float  |
| WindGustDir   | Direction of the strongest gust                       | Compass Points  | object |
| WindGustSpeed | Speed of the strongest gust                           | Kilometers/Hour | object |
| WindDir9am    | Wind direction averaged of 10 minutes prior to 9am    | Compass Points  | object |
| WindDir3pm    | Wind direction averaged of 10 minutes prior to 3pm    | Compass Points  | object |
| WindSpeed9am  | Wind speed averaged of 10 minutes prior to 9am        | Kilometers/Hour | float  |
| WindSpeed3pm  | Wind speed averaged of 10 minutes prior to 3pm        | Kilometers/Hour | float  |
| Humidity9am   | Humidity at 9am                                       | Percent         | float  |
| Humidity3pm   | Humidity at 3pm                                       | Percent         | float  |
| Pressure9am   | Atmospheric pressure reduced to mean sea level at 9am | Hectopascal     | float  |
| Pressure3pm   | Atmospheric pressure reduced to mean sea level at 3pm | Hectopascal     | float  |
| Cloud9am      | Fraction of the sky obscured by cloud at 9am          | Eights          | float  |
| Cloud3pm      | Fraction of the sky obscured by cloud at 3pm          | Eights          | float  |
| Temp9am       | Temperature at 9am                                    | Celsius         | float  |
| Temp3pm       | Temperature at 3pm                                    | Celsius         | float  |
| RainToday     | If there was rain today                               | Yes/No          | object |
| RainTomorrow  | If there is rain tomorrow                             | Yes/No          | float  |

Column definitions were gathered from [http://www.bom.gov.au/climate/dwo/IDCJDW0000.shtml](http://www.bom.gov.au/climate/dwo/IDCJDW0000.shtml?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkML0101ENSkillsNetwork20718538-2022-01-01)



## **Import the required libraries**


In [1]:
# All Libraries required for this lab are listed below. The libraries pre-installed on Skills Network Labs are commented.
# !mamba install -qy pandas==1.3.4 numpy==1.21.4 seaborn==0.9.0 matplotlib==3.5.0 scikit-learn==0.20.1
# Note: If your environment doesn't support "!mamba install", use "!pip install"

In [2]:
# Surpress warnings:
def warn(*args, **kwargs):
    pass
import warnings
warnings.warn = warn

In [3]:
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import LinearRegression
from sklearn import preprocessing
import numpy as np
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn import svm
from sklearn.metrics import jaccard_score
from sklearn.metrics import f1_score
from sklearn.metrics import log_loss
from sklearn.metrics import confusion_matrix, accuracy_score
import sklearn.metrics as metrics

### Importing the Dataset


In [4]:
from pyodide.http import pyfetch

async def download(url, filename):
    response = await pyfetch(url)
    if response.status == 200:
        with open(filename, "wb") as f:
            f.write(await response.bytes())

In [5]:
path='https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-ML0101EN-SkillUp/labs/ML-FinalAssignment/Weather_Data.csv'

In [6]:
await download(path, "Weather_Data.csv")
filename ="Weather_Data.csv"

In [7]:
df = pd.read_csv("Weather_Data.csv")

> Note: This version of the lab is designed for JupyterLite, which necessitates downloading the dataset to the interface. However, when working with the downloaded version of this notebook on your local machines (Jupyter Anaconda), you can simply **skip the steps above of "Importing the Dataset"** and use the URL directly in the `pandas.read_csv()` function. You can uncomment and run the statements in the cell below.


In [8]:
#filepath = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-ML0101EN-SkillUp/labs/ML-FinalAssignment/Weather_Data.csv"
#df = pd.read_csv(filepath)

In [9]:
df.head()

Date  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine WindGustDir  \
0  2/1/2008     19.5     22.4      15.6          6.2       0.0           W   
1  2/2/2008     19.5     25.6       6.0          3.4       2.7           W   
2  2/3/2008     21.6     24.5       6.6          2.4       0.1           W   
3  2/4/2008     20.2     22.8      18.8          2.2       0.0           W   
4  2/5/2008     19.7     25.7      77.4          4.8       0.0           W   

   WindGustSpeed WindDir9am WindDir3pm  ...  Humidity9am  Humidity3pm  \
0             41          S        SSW  ...           92           84   
1             41          W          E  ...           83           73   
2             41        ESE        ESE  ...           88           86   
3             41        NNE          E  ...           83           90   
4             41        NNE          W  ...           88           74   

   Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  \
0       1017.6       1017.4         8         8     20.7     20.9        Yes   
1       1017.9       1016.4         7         7     22.4     24.8        Yes   
2       1016.7       1015.6         7         8     23.5     23.0        Yes   
3       1014.2       1011.8         8         8     21.4     20.9        Yes   
4       1008.3       1004.8         8         8     22.5     25.5        Yes   

   RainTomorrow  
0           Yes  
1           Yes  
2           Yes  
3           Yes  
4           Yes  

[5 rows x 22 columns]

### Data Preprocessing


#### One Hot Encoding


First, we need to perform one hot encoding to convert categorical variables to binary variables.


In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3271 entries, 0 to 3270
Data columns (total 22 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Date           3271 non-null   object 
 1   MinTemp        3271 non-null   float64
 2   MaxTemp        3271 non-null   float64
 3   Rainfall       3271 non-null   float64
 4   Evaporation    3271 non-null   float64
 5   Sunshine       3271 non-null   float64
 6   WindGustDir    3271 non-null   object 
 7   WindGustSpeed  3271 non-null   int64  
 8   WindDir9am     3271 non-null   object 
 9   WindDir3pm     3271 non-null   object 
 10  WindSpeed9am   3271 non-null   int64  
 11  WindSpeed3pm   3271 non-null   int64  
 12  Humidity9am    3271 non-null   int64  
 13  Humidity3pm    3271 non-null   int64  
 14  Pressure9am    3271 non-null   float64
 15  Pressure3pm    3271 non-null   float64
 16  Cloud9am       3271 non-null   int64  
 17  Cloud3pm       3271 non-null   int64  
 18  Temp9am 

In [11]:
df.describe()

MinTemp      MaxTemp     Rainfall  Evaporation     Sunshine  \
count  3271.000000  3271.000000  3271.000000  3271.000000  3271.000000   
mean     14.877102    23.005564     3.342158     5.175787     7.168970   
std       4.554710     4.483752     9.917746     2.757684     3.815966   
min       4.300000    11.700000     0.000000     0.000000     0.000000   
25%      11.000000    19.600000     0.000000     3.200000     4.250000   
50%      14.900000    22.800000     0.000000     4.800000     8.300000   
75%      18.800000    26.000000     1.400000     7.000000    10.200000   
max      27.600000    45.800000   119.400000    18.400000    13.600000   

       WindGustSpeed  WindSpeed9am  WindSpeed3pm  Humidity9am  Humidity3pm  \
count    3271.000000   3271.000000   3271.000000  3271.000000  3271.000000   
mean       41.476307     15.077041     19.294405    68.243962    54.698563   
std        10.806951      7.043825      7.453331    15.086127    16.279241   
min        17.000000      0.000000      0.000000    19.000000    10.000000   
25%        35.000000     11.000000     15.000000    58.000000    44.000000   
50%        41.000000     15.000000     19.000000    69.000000    56.000000   
75%        44.000000     20.000000     24.000000    80.000000    64.000000   
max        96.000000     54.000000     57.000000   100.000000    99.000000   

       Pressure9am  Pressure3pm     Cloud9am     Cloud3pm      Temp9am  \
count  3271.000000  3271.000000  3271.000000  3271.000000  3271.000000   
mean   1018.334424  1016.003085     4.318557     4.176093    17.821461   
std       7.020090     7.019915     2.526923     2.411274     4.894316   
min     986.700000   989.800000     0.000000     0.000000     6.400000   
25%    1013.700000  1011.300000     2.000000     2.000000    13.800000   
50%    1018.600000  1016.300000     5.000000     4.000000    18.200000   
75%    1023.100000  1020.800000     7.000000     7.000000    21.700000   
max    1039.000000  1036.700000     9.000000     8.000000    36.500000   

           Temp3pm  
count  3271.000000  
mean     21.543656  
std       4.297053  
min      10.200000  
25%      18.400000  
50%      21.300000  
75%      24.500000  
max      44.700000

In [12]:
df_sydney_processed = pd.get_dummies(data=df, columns=['RainToday', 'WindGustDir', 'WindDir9am', 'WindDir3pm'])

Next, we replace the values of the 'RainTomorrow' column changing them from a categorical column to a binary column. We do not use the `get_dummies` method because we would end up with two columns for 'RainTomorrow' and we do not want, since 'RainTomorrow' is our target.


In [13]:
df_sydney_processed.replace(['No', 'Yes'], [0,1], inplace=True)

In [14]:
pd.set_option('display.max_columns', None)


In [15]:
df_sydney_processed

Date  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0      2/1/2008     19.5     22.4      15.6          6.2       0.0   
1      2/2/2008     19.5     25.6       6.0          3.4       2.7   
2      2/3/2008     21.6     24.5       6.6          2.4       0.1   
3      2/4/2008     20.2     22.8      18.8          2.2       0.0   
4      2/5/2008     19.7     25.7      77.4          4.8       0.0   
...         ...      ...      ...       ...          ...       ...   
3266  6/21/2017      8.6     19.6       0.0          2.0       7.8   
3267  6/22/2017      9.3     19.2       0.0          2.0       9.2   
3268  6/23/2017      9.4     17.7       0.0          2.4       2.7   
3269  6/24/2017     10.1     19.3       0.0          1.4       9.3   
3270  6/25/2017      7.6     19.3       0.0          3.4       9.4   

      WindGustSpeed  WindSpeed9am  WindSpeed3pm  Humidity9am  Humidity3pm  \
0                41            17            20           92           84   
1                41             9            13           83           73   
2                41            17             2           88           86   
3                41            22            20           83           90   
4                41            11             6           88           74   
...             ...           ...           ...          ...          ...   
3266             37            22            20           73           52   
3267             30            20             7           78           53   
3268             24            15            13           85           56   
3269             43            17            19           56           35   
3270             35            13            13           73           32   

      Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  \
0          1017.6       1017.4         8         8     20.7     20.9   
1          1017.9       1016.4         7         7     22.4     24.8   
2          1016.7       1015.6         7         8     23.5     23.0   
3          1014.2       1011.8         8         8     21.4     20.9   
4          1008.3       1004.8         8         8     22.5     25.5   
...           ...          ...       ...       ...      ...      ...   
3266       1025.9       1025.3         2         2     10.5     17.9   
3267       1028.5       1024.6         2         2     11.0     18.7   
3268       1020.8       1015.0         6         6     10.2     17.3   
3269       1017.3       1015.1         5         2     12.4     19.0   
3270       1018.6       1015.4         1         1      9.4     18.8   

      RainTomorrow  RainToday_No  RainToday_Yes  WindGustDir_E  \
0                1         False           True          False   
1                1         False           True          False   
2                1         False           True          False   
3                1         False           True          False   
4                1         False           True          False   
...            ...           ...            ...            ...   
3266             0          True          False          False   
3267             0          True          False          False   
3268             0          True          False          False   
3269             0          True          False          False   
3270             0          True          False          False   

      WindGustDir_ENE  WindGustDir_ESE  WindGustDir_N  WindGustDir_NE  \
0               False            False          False           False   
1               False            False          False           False   
2               False            False          False           False   
3               False            False          False           False   
4               False            False          False           False   
...               ...              ...            ...             ...   
3266            False            False          False           False   
326

### Training Data and Test Data


Now, we set our 'features' or x values and our Y or target variable.


In [16]:
df_sydney_processed.drop('Date',axis=1,inplace=True)

In [17]:
df_sydney_processed = df_sydney_processed.astype(float)
# converts all trues and falses to 1s and 0s

In [18]:
df_sydney_processed

MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  WindGustSpeed  \
0        19.5     22.4      15.6          6.2       0.0           41.0   
1        19.5     25.6       6.0          3.4       2.7           41.0   
2        21.6     24.5       6.6          2.4       0.1           41.0   
3        20.2     22.8      18.8          2.2       0.0           41.0   
4        19.7     25.7      77.4          4.8       0.0           41.0   
...       ...      ...       ...          ...       ...            ...   
3266      8.6     19.6       0.0          2.0       7.8           37.0   
3267      9.3     19.2       0.0          2.0       9.2           30.0   
3268      9.4     17.7       0.0          2.4       2.7           24.0   
3269     10.1     19.3       0.0          1.4       9.3           43.0   
3270      7.6     19.3       0.0          3.4       9.4           35.0   

      WindSpeed9am  WindSpeed3pm  Humidity9am  Humidity3pm  Pressure9am  \
0             17.0          20.0         92.0         84.0       1017.6   
1              9.0          13.0         83.0         73.0       1017.9   
2             17.0           2.0         88.0         86.0       1016.7   
3             22.0          20.0         83.0         90.0       1014.2   
4             11.0           6.0         88.0         74.0       1008.3   
...            ...           ...          ...          ...          ...   
3266          22.0          20.0         73.0         52.0       1025.9   
3267          20.0           7.0         78.0         53.0       1028.5   
3268          15.0          13.0         85.0         56.0       1020.8   
3269          17.0          19.0         56.0         35.0       1017.3   
3270          13.0          13.0         73.0         32.0       1018.6   

      Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainTomorrow  \
0          1017.4       8.0       8.0     20.7     20.9           1.0   
1          1016.4       7.0       7.0     22.4     24.8           1.0   
2          1015.6       7.0       8.0     23.5     23.0           1.0   
3          1011.8       8.0       8.0     21.4     20.9           1.0   
4          1004.8       8.0       8.0     22.5     25.5           1.0   
...           ...       ...       ...      ...      ...           ...   
3266       1025.3       2.0       2.0     10.5     17.9           0.0   
3267       1024.6       2.0       2.0     11.0     18.7           0.0   
3268       1015.0       6.0       6.0     10.2     17.3           0.0   
3269       1015.1       5.0       2.0     12.4     19.0           0.0   
3270       1015.4       1.0       1.0      9.4     18.8           0.0   

      RainToday_No  RainToday_Yes  WindGustDir_E  WindGustDir_ENE  \
0              0.0            1.0            0.0              0.0   
1              0.0            1.0            0.0              0.0   
2              0.0            1.0            0.0              0.0   
3              0.0            1.0            0.0              0.0   
4              0.0            1.0            0.0              0.0   
...            ...            ...            ...              ...   
3266           1.0            0.0            0.0              0.0   
3267           1.0            0.0            0.0              0.0   
3268           1.0            0.0            0.0              0.0   
3269           1.0            0.0            0.0              0.0   
3270           1.0            0.0            0.0              0.0   

      WindGustDir_ESE  WindGustDir_N  WindGustDir_NE  WindGustDir_NNE  \
0                 0.0            0.0             0.0              0.0   
1                 0.0            0.0             0.0              0.0   
2                 0.0            0.0             0.0              0.0   
3                 0.0            0.0             0.0              0.0   
4                 0.0            0.0             0.0              0.0   
...               ...            ...             ...              ...   

In [19]:
features = df_sydney_processed.drop(columns='RainTomorrow', axis=1)
Y = df_sydney_processed['RainTomorrow']

In [20]:
features

MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  WindGustSpeed  \
0        19.5     22.4      15.6          6.2       0.0           41.0   
1        19.5     25.6       6.0          3.4       2.7           41.0   
2        21.6     24.5       6.6          2.4       0.1           41.0   
3        20.2     22.8      18.8          2.2       0.0           41.0   
4        19.7     25.7      77.4          4.8       0.0           41.0   
...       ...      ...       ...          ...       ...            ...   
3266      8.6     19.6       0.0          2.0       7.8           37.0   
3267      9.3     19.2       0.0          2.0       9.2           30.0   
3268      9.4     17.7       0.0          2.4       2.7           24.0   
3269     10.1     19.3       0.0          1.4       9.3           43.0   
3270      7.6     19.3       0.0          3.4       9.4           35.0   

      WindSpeed9am  WindSpeed3pm  Humidity9am  Humidity3pm  Pressure9am  \
0             17.0          20.0         92.0         84.0       1017.6   
1              9.0          13.0         83.0         73.0       1017.9   
2             17.0           2.0         88.0         86.0       1016.7   
3             22.0          20.0         83.0         90.0       1014.2   
4             11.0           6.0         88.0         74.0       1008.3   
...            ...           ...          ...          ...          ...   
3266          22.0          20.0         73.0         52.0       1025.9   
3267          20.0           7.0         78.0         53.0       1028.5   
3268          15.0          13.0         85.0         56.0       1020.8   
3269          17.0          19.0         56.0         35.0       1017.3   
3270          13.0          13.0         73.0         32.0       1018.6   

      Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday_No  \
0          1017.4       8.0       8.0     20.7     20.9           0.0   
1          1016.4       7.0       7.0     22.4     24.8           0.0   
2          1015.6       7.0       8.0     23.5     23.0           0.0   
3          1011.8       8.0       8.0     21.4     20.9           0.0   
4          1004.8       8.0       8.0     22.5     25.5           0.0   
...           ...       ...       ...      ...      ...           ...   
3266       1025.3       2.0       2.0     10.5     17.9           1.0   
3267       1024.6       2.0       2.0     11.0     18.7           1.0   
3268       1015.0       6.0       6.0     10.2     17.3           1.0   
3269       1015.1       5.0       2.0     12.4     19.0           1.0   
3270       1015.4       1.0       1.0      9.4     18.8           1.0   

      RainToday_Yes  WindGustDir_E  WindGustDir_ENE  WindGustDir_ESE  \
0               1.0            0.0              0.0              0.0   
1               1.0            0.0              0.0              0.0   
2               1.0            0.0              0.0              0.0   
3               1.0            0.0              0.0              0.0   
4               1.0            0.0              0.0              0.0   
...             ...            ...              ...              ...   
3266            0.0            0.0              0.0              0.0   
3267            0.0            0.0              0.0              0.0   
3268            0.0            0.0              0.0              0.0   
3269            0.0            0.0              0.0              0.0   
3270            0.0            0.0              0.0              0.0   

      WindGustDir_N  WindGustDir_NE  WindGustDir_NNE  WindGustDir_NNW  \
0               0.0             0.0              0.0              0.0   
1               0.0             0.0              0.0              0.0   
2               0.0             0.0              0.0              0.0   
3               0.0             0.0              0.0              0.0   
4               0.0             0.0              0.0              0.0   
...             ...             ... 

### Linear Regression


Understanding Relationships Between Variables
Feature Impact Analysis: Linear regression can help identify the relationship between weather variables and the target variable. For instance, by fitting a linear regression model, one can assess the contribution or influence of factors like "Humidity9am" or "Pressure3pm" on "RainTomorrow." The coefficients of the linear model provide insights into how much change in the predictors (independent variables) impacts the dependent variable.


### Prediction of Continuous Variables

In the context of weather forecasting, continuous variables are those that can take on any value within a certain range and are not restricted to specific categories or discrete outcomes. Examples of continuous variables include temperature, rainfall amount, wind speed, and humidity. These variables can vary smoothly and can have fractional values, making them ideal for prediction using models like linear regression.

### **How Continuous Variable Prediction Works:**

#### 1. **Understanding Continuous Variables:**
   - **Temperature**: This is a continuous variable because it can be any value, such as 20.5°C or 21.7°C, not just specific, fixed categories.
   - **Rainfall**: The amount of rainfall is measured in millimeters and can be any value from 0.0 mm to several hundred millimeters.
   - **Wind Speed**: Wind speed is usually measured in kilometers per hour (km/h) and can range from 0 km/h to very high values, depending on the weather conditions.

#### 2. **Linear Regression Model:**
   - **What It Does**: Linear regression models the relationship between one or more independent variables (also known as features or predictors) and a dependent variable (the target). The model assumes a linear relationship, meaning the change in the dependent variable is proportional to the change in the independent variables.
   - **Example**: Suppose you want to predict the amount of rainfall (`Rainfall`) based on other weather metrics like `Humidity9am`, `Pressure9am`, and `Temp9am`. The linear regression model would take these predictors and attempt to fit a line (or hyperplane in multiple dimensions) that best represents the relationship between these predictors and the target (rainfall).

#### 3. **Model Output:**
   - The output of a linear regression model is a continuous value, which in this case would be the predicted amount of rainfall. 
   - For example, the model might predict that given a certain temperature, humidity, and pressure, the expected rainfall would be 12.4 mm. This is a continuous value because it can take on any numerical value within the plausible range.

#### 4. **Why Linear Regression is Appropriate:**
   - **Nature of Data**: Since the target variable (e.g., rainfall amount) is continuous, linear regression is suitable because it is designed to predict such continuous outcomes.
   - **Prediction Process**: The model works by minimizing the difference between the actual observed values and the predicted values (the residuals) through a process called "least squares," which adjusts the model parameters to fit the data as closely as possible.

### **Example Scenario:**

Let's say you have a dataset where you've recorded the daily temperature, humidity, and atmospheric pressure over several years. You want to predict how much rainfall will occur on a given day based on that day's temperature, humidity, and pressure.

- **Independent Variables (Predictors):** Temperature, Humidity, Pressure.
- **Dependent Variable (Target):** Rainfall (continuous).

Using linear regression, you create a model that finds the best-fitting linear relationship between these variables and the rainfall amount. Once trained, this model can predict the expected rainfall on a new day given the temperature, humidity, and pressure for that day.

### **Mathematical Representation:**

The linear regression model can be represented as:

\[ \text{Rainfall} = \beta_0 + \beta_1 \times \text{Temp9am} + \beta_2 \times \text{Humidity9am} + \beta_3 \times \text{Pressure9am} + \epsilon \]

Where:
- \( \beta_0 \) is the intercept,
- \( \beta_1, \beta_2, \beta_3 \) are the coefficients representing the impact of each predictor,
- \( \epsilon \) is the error term (residuals).

The goal is to find the values of \( \beta_0, \beta_1, \beta_2, \beta_3 \) that minimize the sum of squared residuals, thereby providing the best prediction of rainfall based on the given inputs.

### Conclusion:
In summary, linear regression is ideal for predicting continuous weather-related outcomes like temperature, rainfall, or wind speed because it can model and output continuous values, which can vary smoothly over a range. This makes it a powerful tool for forecasting and analyzing continuous weather metrics.

#### Q1) Use the `train_test_split` function to split the `features` and `Y` dataframes with a `test_size` of `0.2` and the `random_state` set to `10`.


In [21]:
#Enter Your Code and Execute

X_train, X_test, y_train, y_test = train_test_split(features, Y, test_size=0.2, random_state=10, stratify=Y)


#### Q2) Create and train a Linear Regression model called LinearReg using the training data (`x_train`, `y_train`).


In [22]:
#Enter Your Code and Execute

In [23]:
LinearReg = LinearRegression()

In [24]:
LinearReg.fit(X_train,y_train)

LinearRegression()

#### Q3) Now use the `predict` method on the testing data (`x_test`) and save it to the array `predictions`.


In [25]:
X_train

MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  WindGustSpeed  \
1367     20.8     27.5      31.6          3.0      12.5           41.0   
706      19.0     25.1       0.2          3.8       3.0           41.0   
1580     11.1     22.2       0.0          5.8       8.3           76.0   
3064     18.2     25.2       0.0          6.4       4.4           50.0   
1768     14.4     28.7       0.0          4.6      10.1           28.0   
...       ...      ...       ...          ...       ...            ...   
737      20.2     30.0      15.2          2.8       9.9           41.0   
2636     12.4     22.5       0.2          1.6      10.4           35.0   
610      11.5     19.6       6.6          2.6       6.4           41.0   
139      13.8     17.9       3.6          0.6       0.6           41.0   
1606     14.2     20.9       0.0          5.4       6.0           39.0   

      WindSpeed9am  WindSpeed3pm  Humidity9am  Humidity3pm  Pressure9am  \
1367          22.0          24.0         79.0         59.0       1006.8   
706            9.0          17.0         83.0         74.0       1018.9   
1580          19.0          37.0         36.0         35.0       1007.7   
3064           6.0          31.0         61.0         57.0       1012.4   
1768           7.0          15.0         59.0         29.0       1019.0   
...            ...           ...          ...          ...          ...   
737           11.0          17.0         95.0         67.0       1006.7   
2636          15.0          15.0         74.0         60.0       1019.9   
610           22.0          17.0         77.0         64.0       1020.6   
139           19.0          19.0         81.0         60.0       1018.3   
1606          19.0          26.0         71.0         61.0       1019.4   

      Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday_No  \
1367       1005.7       5.0       4.0     22.4     25.8           0.0   
706        1017.2       8.0       7.0     21.7     24.4           1.0   
1580       1004.3       1.0       6.0     15.9     14.5           1.0   
3064       1008.8       7.0       2.0     22.4     24.2           1.0   
1768       1016.0       1.0       1.0     18.6     28.1           1.0   
...           ...       ...       ...      ...      ...           ...   
737        1003.5       6.0       3.0     21.4     28.5           0.0   
2636       1015.2       1.0       2.0     16.3     21.3           1.0   
610        1016.7       6.0       4.0     13.5     18.0           0.0   
139        1016.2       7.0       7.0     14.3     17.2           0.0   
1606       1016.9       7.0       6.0     16.4     19.4           1.0   

      RainToday_Yes  WindGustDir_E  WindGustDir_ENE  WindGustDir_ESE  \
1367            1.0            0.0              0.0              0.0   
706             0.0            0.0              0.0              0.0   
1580            0.0            0.0              0.0              0.0   
3064            0.0            0.0              0.0              0.0   
1768            0.0            0.0              0.0              0.0   
...             ...            ...              ...              ...   
737             1.0            0.0              0.0              0.0   
2636            0.0            0.0              0.0              0.0   
610             1.0            0.0              0.0              0.0   
139             1.0            0.0              0.0              0.0   
1606            0.0            0.0              0.0              0.0   

      WindGustDir_N  WindGustDir_NE  WindGustDir_NNE  WindGustDir_NNW  \
1367            0.0             0.0              0.0              0.0   
706             0.0             0.0              0.0              0.0   
1580            0.0             0.0              0.0              0.0   
3064            0.0             1.0              0.0              0.0   
1768            0.0             0.0              0.0              0.0   
...             ...             ... 

In [26]:
#Enter Your Code and Execute

In [27]:
y_train

1367    0.0
706     0.0
1580    0.0
3064    0.0
1768    0.0
       ... 
737     0.0
2636    0.0
610     1.0
139     1.0
1606    0.0
Name: RainTomorrow, Length: 2616, dtype: float64

In [28]:
predictions = LinearReg.predict(X_test)

In [29]:
predictions

array([-2.04023719e-02,  1.73963189e-01,  4.06603634e-01,  9.37716961e-02,
        1.68298423e-01,  1.65078521e-01,  4.31520164e-01, -4.72609401e-02,
        2.91781485e-01, -4.22515869e-02, -5.78672290e-02,  1.61349773e-01,
       -3.50881219e-02,  3.10630560e-01,  2.33495116e-01,  1.46781266e-01,
        2.01474011e-01,  4.03541923e-02,  3.89600575e-01, -4.35812473e-02,
        2.74687231e-01,  1.46573901e-01,  2.17129230e-01,  5.68175018e-01,
        4.40535545e-01,  4.09880280e-02,  2.16342390e-01, -7.49880075e-02,
        7.78726459e-01, -3.86001468e-02,  1.04065824e+00,  7.42452502e-01,
        1.56271577e-01, -4.56249714e-03,  4.07938123e-01,  2.79417276e-01,
        3.59704435e-01, -3.64722013e-02, -8.88547301e-02,  6.05404973e-02,
        8.21586311e-01,  6.05604231e-01,  1.56773031e-01,  2.25477338e-01,
        3.82341444e-01,  6.10482633e-01,  2.16909468e-01,  2.20777571e-01,
        3.33259225e-01,  1.68810129e-01, -6.01975322e-02, -9.60844159e-02,
        1.85928166e-01,  

In [30]:
import numpy as np

# Assuming predictions are the continuous outputs from Linear Regression
# Convert predictions to binary using a threshold of 0.5
binary_predictions = (predictions > 0.5).astype(int)

# Print the binary predictions
print(binary_predictions)


[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1 0 1 1 0 0 0 0 0
 0 0 0 1 1 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 0 1 0 0 0 0 0 0 0 1 1 1
 0 1 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 0 1 0 1 0 0 0 1 0 0 0 0 1 0 0 1 0 0 1
 0 0 1 0 0 1 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 1 0 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0 0 1 1 0 0 0 0 0 1 1 0 0 0 1 1 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1 0 1 0 0 1 0 0 0 0 1 1 1 0 0 1
 0 0 0 0 0 0 0 0 0 1 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 0 0 1 1 0 0 0 0
 0 0 0 0 1 1 0 0 0 1 1 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 1 0 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 0
 1 0 0 1 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 0 1 1 0 1 0 0 1 1 0 0 0
 0 0 0 0 0 0 1 0 0 0 0 0 0 1 0 1 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0 0 0 0 0 1 0
 0 0 0 1 0 0 0 0 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 1 0 1 0 0 0 1 1 0 0 0 0
 0 0 0 0 0 0 1 0 0 0 1 0 0 0 0 0 0 1 0 1 1 1 0 0 1 0 0 0 0 0 1 0 0 0 0 1 0
 0 1 0 0 0 1 1 0 0 1 0 0 

#### Q4) Using the `predictions` and the `y_test` dataframe calculate the value for each metric using the appropriate function.


In [31]:
#Enter Your Code and Execute

In [32]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score


In [33]:
LinearRegression_MAE = mean_absolute_error(y_test, predictions)
LinearRegression_MSE = mean_squared_error(y_test, predictions)
LinearRegression_R2 = r2_score(y_test, predictions)

#### Q5) Show the MAE, MSE, and R2 in a tabular format using data frame for the linear model.


In [34]:
#Enter Your Code and Execute

In [35]:
Report  = pd.DataFrame({
    'Metric': ['Mean Absolute Error (MAE)', 'Mean Squared Error (MSE)', 'R-squared'],
    'Value': [LinearRegression_MAE, LinearRegression_MSE, LinearRegression_R2]
})

In [36]:
Report

Metric     Value
0  Mean Absolute Error (MAE)  0.261357
1   Mean Squared Error (MSE)  0.127618
2                  R-squared  0.335944

In [37]:
df['Date']

0        2/1/2008
1        2/2/2008
2        2/3/2008
3        2/4/2008
4        2/5/2008
          ...    
3266    6/21/2017
3267    6/22/2017
3268    6/23/2017
3269    6/24/2017
3270    6/25/2017
Name: Date, Length: 3271, dtype: object

In [38]:
y_test

2273    0.0
346     0.0
528     1.0
505     0.0
2590    0.0
       ... 
1498    0.0
1645    0.0
2432    1.0
2188    0.0
1881    0.0
Name: RainTomorrow, Length: 655, dtype: float64

In [39]:
# y_test['prediction'] = LinearReg.predict(X_test) # Make sure to create the 'prediction' column in 'test_11'


In [40]:
# df.columns

### KNN


## KNN Explained in the easiest way possible: 

Imagine you’re at a new school and you’re trying to figure out who your new friends might be. One way to do that is by looking at what kind of things other kids like—like if they’re into the same games or subjects as you. If you find kids who like the same things, you might guess they could be your friends, right?

That's kind of how K-Nearest Neighbors (KNN) works.

Here’s the idea:
You're the new kid: In the world of KNN, you're like a "new point," and we want to figure out what "group" or "category" you belong to (like "friends who love soccer" or "friends who love video games").
Find kids who are similar to you: KNN works by looking at a group of other points (or kids) that already have their "labels" (we know if they love soccer or video games). Then, it finds the ones closest to you. Just like you'd look for kids who like the same things, KNN finds the points that are most similar.
K stands for how many friends you look at: The "K" in KNN is the number of neighbors (or similar kids) you check. If K = 3, you’ll look at the 3 closest kids to see what they like.
Decide your group: Once you've looked at your nearest neighbors, you decide which group you fit into based on what most of those kids like. So, if 2 of the 3 closest kids love soccer, you'll probably be part of the soccer group too!
Summary:
KNN is like trying to find out who your friends are by checking out who’s most similar to you.
The "K" tells you how many kids (or points) to compare yourself with.
You then join the group that most of your nearest neighbors belong to.


Alright, let's tie the "you're the new kid" example to the idea of KNN being a lazy learning algorithm!

Remember the scenario:
You’re the new kid at school, and you’re trying to figure out which group of friends you should hang out with based on your interests. Now let’s think about how this works with KNN being lazy.

How KNN is lazy in this context:
Before you show up (training phase): The school doesn’t bother figuring out which groups you’ll fit into in advance. No one has pre-organized the students into specific groups for you. They don’t make a "friend-making plan" beforehand.
When you arrive (prediction phase): The moment you walk into school, that’s when everyone starts paying attention. Instead of having groups ready, you (and KNN) now do the work of figuring out where you belong.
You start by comparing yourself to all the kids around you. You look at what they’re into (sports, games, etc.).
You find a few kids who like similar things (like soccer or video games), and you ask yourself, “Who’s most like me?”
Decision-making: Based on which kids you found are closest to you in interests, you join the group that has the most kids with similar interests. If most of your closest new friends like soccer, you join the soccer group!
Why is this "lazy"?
KNN is lazy because it doesn’t make any decisions or organize anything until you (the new kid) actually arrive. It waits until the last moment, then does all the comparing and decision-making right when it’s needed. Other algorithms might plan ahead and organize all the kids into groups before you even show up, but KNN is more like, "I'll figure it out when the time comes!"

So, just like the school doesn’t plan where you’ll fit until you arrive, KNN waits until new data (you) appears before it does the work of comparing and deciding. That's why it's called a lazy learning algorithm!

Let’s relate the K-Nearest Neighbors (KNN) concept from the "new kid at school" example to our weatherAUS.csv dataset:

KNN with Weather Data: Mapping the "New Kid" Example
The New Kid:
In the context of our weather dataset, the "new kid" is a new day's weather data that you want to predict or classify. For instance, you might have a new day's weather observations, and you want to predict whether it will rain tomorrow based on past data.
Training Data (Old Kids):
The training data consists of weather records from the past (2008 to 2017). Each record (day) has features like temperature, rainfall, humidity, etc., and a label (whether it rained the next day or not).
Finding Neighbors:
To make a prediction about the new day’s weather, you compare it to all the past weather records (old kids). You look for the days that have similar weather conditions to the new day. This involves comparing features like temperature, humidity, and rainfall.
Making Predictions:
After finding the closest days (neighbors) based on weather metrics, you see what happened on those similar days (e.g., whether it rained or not). If most of the similar days had rain, you predict that it will likely rain on the new day as well.
Practical Steps with Weather Data:
Preparing Data:
Convert the weather data into a format suitable for KNN. This includes handling missing values, converting categorical features (like RainToday and RainTomorrow) into numerical values, and scaling features if needed.
Comparing Data:
Use metrics like Euclidean distance to find the closest days to your new day. For example, if today’s temperature and humidity are close to those of past days, those past days become the “neighbors” you consider.
Making a Decision:
Based on the majority class of the neighbors (e.g., whether it rained or not), make your prediction for the new day. If most of the nearest neighbors had rain tomorrow, predict rain.


## Summary:
New Kid: New day's weather data.
Old Kids: Historical weather data.
Finding Neighbors: Comparing new day's weather to past days to find similar conditions.
Prediction: Based on what happened on similar past days (e.g., whether it rained), predict the weather for the new day.
In this way, the KNN algorithm uses past data (historical weather records) to make predictions about new data (the new day’s weather).

#### Q6) Create and train a KNN model called KNN using the training data (`x_train`, `y_train`) with the `n_neighbors` parameter set to `4`.


In [41]:
#Enter Your Code and Execute

In [42]:
KNN = KNeighborsClassifier(n_neighbors=4)

In [43]:
KNN.fit(X_train,y_train)

KNeighborsClassifier(n_neighbors=4)

#### Q7) Now use the `predict` method on the testing data (`x_test`) and save it to the array `predictions`.


In [44]:
#Enter Your Code and Execute

In [45]:
predictions = KNN.predict(X_test)

In [46]:
y_test

2273    0.0
346     0.0
528     1.0
505     0.0
2590    0.0
       ... 
1498    0.0
1645    0.0
2432    1.0
2188    0.0
1881    0.0
Name: RainTomorrow, Length: 655, dtype: float64

In [47]:
predictions

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 1., 1., 0., 0.,
       1., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 0., 0., 1., 0., 1., 0., 0.,
       0., 0., 0., 1., 1., 0., 0., 0., 0., 0., 1., 0., 1., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 1., 0., 0., 0., 1., 0., 0.,
       0., 0., 1., 0., 0., 1., 0., 0., 1., 0., 0., 1., 0., 0., 1., 0., 0.,
       1., 0., 0., 0., 0., 0., 1., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 0., 0.,
       0., 0., 0., 1., 1., 0., 0., 0., 1., 1., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 1., 0., 0.,
       0., 0., 1., 0., 1., 0., 0., 1., 0., 0., 0., 0., 1., 1., 0., 0., 0.,
       0., 0., 1., 0., 1.

#### Q8) Using the `predictions` and the `y_test` dataframe calculate the value for each metric using the appropriate function.


In [48]:
#Enter Your Code and Execute
from sklearn.metrics import accuracy_score

In [49]:
print(f"Length of y_test: {len(y_test)}")
print(f"Length of predictions: {len(predictions)}")


Length of y_test: 655
Length of predictions: 655


In [50]:
KNN_Accuracy_Score = accuracy_score(y_test, predictions)
KNN_JaccardIndex = jaccard_score(y_test, predictions)
KNN_F1_Score = f1_score(y_test, predictions)

In [51]:
KNN_report=pd.DataFrame({
     'Metric': ['KNN_Accuracy_Score', 'KNN_JaccardIndex', 'KNN_F1_Score'],
    'Value': [KNN_Accuracy_Score, KNN_JaccardIndex, KNN_F1_Score]
})

Accuracy Score: The ratio of correctly predicted instances to the total instances.
--------
Jaccard Index: Measures the similarity between predicted and actual labels, typically used for binary or multilabel classification.
----------
F1 Score: The harmonic mean of precision and recall, balancing the two.
------------

In [52]:
KNN_report

Metric     Value
0  KNN_Accuracy_Score  0.815267
1    KNN_JaccardIndex  0.366492
2        KNN_F1_Score  0.536398

 Here's a brief explanation of each metric and its value:
### 1. **KNN Accuracy Score: 0.815267**
   - **Explanation**: This indicates that **81.53%** of the predictions made by the KNN model are correct. The **accuracy score** is the ratio of correctly predicted instances to the total instances.
   - **Formula**:
     \[
     \text{Accuracy} = \frac{\text{Number of Correct Predictions}}{\text{Total Number of Predictions}}
     \]
### 2. **KNN Jaccard Index: 0.366492**
   - **Explanation**: The **Jaccard Index** (also known as the **Intersection over Union**) measures similarity between the predicted and actual labels. A Jaccard Index of **0.366** means that 36.65% of the predicted and actual labels overlap.
   - **Formula**:
     \[
     \text{Jaccard Index} = \frac{| \text{True Positives} |}{| \text{True Positives} + \text{False Positives} + \text{False Negatives} |}
     \]
### 3. **KNN F1 Score: 0.536398**
   - **Explanation**: The **F1 Score** is the harmonic mean of precision and recall, and it's useful for imbalanced datasets. A value of **0.536** (53.64%) indicates a balance between precision and recall, where both are considered equally important.
   - **Formula**:
     \[
     \text{F1 Score} = 2 \cdot \frac{\text{Precision} \cdot \text{Recall}}{\text{Precision} + \text{Recall}}
     \]
   - Where:
     - **Precision**: Measures the correctness of positive predictions.
     - **Recall**: Measures the ability to find all positive instances.
### Summary of Metrics:
- **Accuracy Score (81.53%)**: Indicates overall correctness of predictions.
- **Jaccard Index (36.65%)**: Shows overlap between actual and predicted labels.
- **F1 Score (53.64%)**: Balances precision and recall, useful for imbalanced data.

### 1. **Jaccard Index (36.65%)**
   - Imagine you have two circles. One circle represents all the days your KNN model **predicted it would rain**. The other circle represents the days it **actually rained**.
   - The **Jaccard Index** measures how much these two circles overlap. It looks at how many days both circles agree on (rain predicted and it rained), and compares that to the total number of days in both circles.
   **Example**:
   - If you predicted rain on 10 days, and it actually rained on 8 of those days, the Jaccard Index shows how good your predictions are at matching the reality. A score of **36.65%** means that only about one-third of your predictions and actual results match.
   **Simple Formula**:
   - Jaccard Index = \(\frac{\text{Days you got right}}{\text{Days you predicted rain + Days it actually rained} - \text{Days you got right}}\)
---
### 2. **F1 Score (53.64%)**
   - The **F1 Score** is like grading yourself on both how **accurate** you are and how good you are at **finding all the rainy days**. It combines two things:
     1. How many of your rain predictions were **correct**.
     2. How many days it actually rained that you **caught** in your predictions.
   - Think of it like playing a game of guessing: you're trying to guess which days it will rain, and the F1 Score checks both how **careful** and how **thorough** your guesses are.
   **Example**:
   - If you guessed that it would rain on 10 days, but it only rained on 5 of those, the F1 Score checks how well you guessed. A score of **53.64%** means you're doing okay—you're getting some rainy days right, but there's room for improvement in guessing both more carefully and more thoroughly.
   **Simple Formula**:
   - F1 Score = \(\frac{2 \times \text{correct guesses}}{\text{total guesses + all rainy days}}\)
---

- **Jaccard Index** tells you how much your predictions match what really happened, like how much two circles overlap.
- **F1 Score** is like a report card that grades you on how well you're guessing and catching the rainy days, considering both your **accuracy** and how many rainy days you missed.


# Decision Tree


A decision tree is a popular machine learning model used for both classification and regression tasks. It works by recursively splitting the dataset into smaller subsets based on certain conditions, eventually arriving at a decision or prediction.

### Key Concepts in a Decision Tree:
1. **Root Node**: 
   - The starting point of the tree that represents the entire dataset. It contains the first condition or feature to split on.

2. **Internal Nodes**:
   - These are the nodes where the data gets split based on a condition (e.g., whether a feature is greater than or less than a certain value). Each internal node represents a decision based on one feature.

3. **Branches**: 
   - These are the outcomes of a decision (split) at an internal node. Each branch represents the result of the condition (true/false or a range of values).

4. **Leaf Nodes (Terminal Nodes)**:
   - These nodes represent the final outcome or decision, whether it's a class label (in classification) or a continuous value (in regression).

### How a Decision Tree Works:
- **Recursive Splitting**: 
  The tree starts at the root node and splits the dataset based on the most important feature (determined by a metric like Gini Index, Information Gain, or others). The process continues recursively for each child node until one of the following conditions is met:
  - All data in a node belongs to the same class (in classification tasks).
  - A maximum tree depth is reached (to avoid overfitting).
  - Further splits do not add value (measured by a stopping criterion like minimum samples required to split).

### Splitting Criteria:
- **Gini Impurity (for classification)**: 
  Measures how "pure" a node is. A pure node has instances that belong to only one class. The algorithm aims to reduce the Gini impurity with each split.
  
- **Information Gain (Entropy)**: 
  Measures the reduction in uncertainty (or entropy) after splitting the data based on a feature. The algorithm seeks to maximize information gain.

- **Mean Squared Error (for regression)**: 
  Used in regression trees to split the data in a way that minimizes the variance of the target variable.

### Example (Classification Tree):
Imagine you’re trying to classify whether a person buys a house based on income and age. The root node might check the income level, splitting the dataset into "high income" and "low income." Each branch might then split again based on age, and so on. Eventually, the leaf nodes will predict whether the person buys a house or not.

### Advantages:
1. **Easy to Understand and Interpret**: Decision trees resemble human decision-making.
2. **Non-Parametric**: They don’t assume any particular distribution of the data.
3. **Feature Selection**: Automatically selects important features during splitting.

### Disadvantages:
1. **Overfitting**: Trees can become too complex and overfit the training data, especially if they are not pruned.
2. **Instability**: Small changes in the data can result in entirely different splits.
3. **Biased towards features with more levels**: Features with many possible values (like continuous variables) might dominate the tree splits.

### Regularization Techniques:
1. **Pruning**: 
   - Involves removing unnecessary branches that add little predictive power. This reduces overfitting and simplifies the model.
   
2. **Max Depth**: 
   - Limits how deep the tree can go, controlling its complexity.
   
3. **Min Samples per Split**: 
   - Sets a minimum number of samples required to further split a node.

### Visualization:
Decision trees can be visualized easily, where each decision node is a condition on a feature, and branches represent outcomes, making it highly interpretable.


#### Q9) Create and train a Decision Tree model called Tree using the training data (`x_train`, `y_train`).


In [53]:
#Enter Your Code and Execute

In [54]:
Tree = DecisionTreeClassifier(random_state=42)

In [55]:
Tree.fit(X_train,y_train)

DecisionTreeClassifier(random_state=42)

#### Q10) Now use the `predict` method on the testing data (`x_test`) and save it to the array `predictions`.


In [56]:
#Enter Your Code and Execute

In [57]:
predictions = Tree.predict(X_test)

In [58]:
predictions

array([0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 1., 0., 0., 1., 1., 0., 1., 0., 0., 0., 1., 1., 1., 0.,
       1., 1., 0., 0., 0., 0., 1., 1., 0., 0., 0., 1., 1., 0., 0., 0., 0.,
       0., 0., 0., 1., 0., 0., 1., 0., 1., 0., 0., 0., 1., 0., 0., 0., 0.,
       0., 0., 0., 1., 1., 1., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 1.,
       0., 0., 0., 0., 0., 0., 0., 1., 1., 0., 1., 0., 0., 0., 1., 0., 0.,
       0., 0., 1., 1., 0., 1., 0., 0., 1., 0., 0., 1., 0., 0., 1., 0., 0.,
       0., 0., 0., 1., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0.,
       0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1.,
       0., 1., 0., 0., 0., 0., 0., 1., 0., 1., 0., 1., 0., 0., 1., 0., 0.,
       0., 0., 0., 1., 1., 1., 0., 0., 1., 1., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 1.,
       1., 0., 1., 1., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 1., 0., 0.,
       1., 0., 1., 0., 1.

In [59]:
# Import necessary libraries
from sklearn.tree import DecisionTreeClassifier

# Create and train the model (as already done)
Tree = DecisionTreeClassifier(criterion="entropy", random_state=42)
Tree.fit(X_train, y_train)

# Access the root node information
tree_structure = Tree.tree_

features_1 = [
    "MinTemp", "MaxTemp", "Rainfall", "Evaporation", "Sunshine", "WindGustSpeed",
    "WindSpeed9am", "WindSpeed3pm", "Humidity9am", "Humidity3pm", "Pressure9am", 
    "Pressure3pm", "Cloud9am", "Cloud3pm", "Temp9am", "Temp3pm", "RainToday_No", 
    "RainToday_Yes", "WindGustDir_E", "WindGustDir_ENE", "WindGustDir_ESE", 
    "WindGustDir_N", "WindGustDir_NE", "WindGustDir_NNE", "WindGustDir_NNW", 
    "WindGustDir_NW", "WindGustDir_S", "WindGustDir_SE", "WindGustDir_SSE", 
    "WindGustDir_SSW", "WindGustDir_SW", "WindGustDir_W", "WindGustDir_WNW", 
    "WindGustDir_WSW", "WindDir9am_E", "WindDir9am_ENE", "WindDir9am_ESE", 
    "WindDir9am_N", "WindDir9am_NE", "WindDir9am_NNE", "WindDir9am_NNW", 
    "WindDir9am_NW", "WindDir9am_S", "WindDir9am_SE", "WindDir9am_SSE", 
    "WindDir9am_SSW", "WindDir9am_SW", "WindDir9am_W", "WindDir9am_WNW", 
    "WindDir9am_WSW", "WindDir3pm_E", "WindDir3pm_ENE", "WindDir3pm_ESE", 
    "WindDir3pm_N", "WindDir3pm_NE", "WindDir3pm_NNE", "WindDir3pm_NNW", 
    "WindDir3pm_NW", "WindDir3pm_S", "WindDir3pm_SE", "WindDir3pm_SSE", 
    "WindDir3pm_SSW", "WindDir3pm_SW", "WindDir3pm_W", "WindDir3pm_WNW", 
    "WindDir3pm_WSW"
]
# The feature that the root node splits on
root_feature_index = tree_structure.feature[0]  # The feature index of the root node
root_threshold = tree_structure.threshold[0]    # The threshold value of the root node

# Print the feature and threshold
print(f"Root node splits on feature index: {root_feature_index}")
print(f"Root node threshold: {root_threshold}")

# If you have feature names, you can print the actual feature name
feature_names = features_1  # Replace with actual feature names
print(f"Root node splits on feature: {feature_names[root_feature_index]}")


Root node splits on feature index: 9
Root node threshold: 73.5
Root node splits on feature: Humidity3pm


Entropy in decision trees is a measure of the **impurity** or **uncertainty** in a dataset. It is used to determine how to split the data at each node in a decision tree. The goal of a decision tree is to create splits that reduce uncertainty and lead to more homogeneous or "pure" subsets of data. In this context, entropy helps quantify how well a split improves the classification of the data.

### Key Concepts:

### Intuition Behind Entropy:
- **High Entropy**: A high entropy value (close to 1) indicates that the dataset is **impure**, meaning the classes are well-mixed. For example, if 50% of the examples belong to one class and 50% to another, the entropy is 1 (maximum uncertainty).
  
- **Low Entropy**: A low entropy value (close to 0) indicates that the dataset is **pure**, meaning the majority of the examples belong to one class. If all examples belong to the same class, entropy is 0 (no uncertainty).

### Example of Entropy:

- If a dataset has 8 samples: 4 are "Yes" and 4 are "No":
  
  \[
  H = - (0.5 \log_2 0.5 + 0.5 \log_2 0.5) = 1
  \]
  This indicates high uncertainty.

- If the dataset is 7 "Yes" and 1 "No":
  
  \[
  H = - (0.875 \log_2 0.875 + 0.125 \log_2 0.125) \approx 0.543
  \]
  This indicates lower uncertainty, as the majority belongs to one class.

### Entropy in Decision Trees:

1. **Splitting Criteria**:
   In decision trees, the algorithm looks for the split that **minimizes entropy**. A good split will result in subsets of data that are as pure as possible (low entropy). Each split aims to reduce the entropy of the target variable.

2. **Information Gain**:
   When building a decision tree, the key criterion is **information gain**. It measures how much uncertainty (entropy) is reduced by a particular split. The split with the highest information gain is chosen.

### Example of Entropy Calculation in a Decision Tree:

Imagine a dataset where you're predicting if it will rain based on "Humidity." You split the data into two groups: "High Humidity" and "Low Humidity."

- Before the split, there are 50% rainy days and 50% non-rainy days (high entropy, \( H = 1 \)).
- After the split:
  - In the "High Humidity" group, 80% of days are rainy (lower entropy, \( H \approx 0.72 \)).
  - In the "Low Humidity" group, 30% of days are rainy (also lower entropy, \( H \approx 0.88 \)).

The decision tree algorithm will choose the split (on humidity) that reduces the overall entropy, creating purer subsets of data.

### Conclusion:

- **Entropy** in decision trees measures the **uncertainty** of data.
- The tree-building algorithm tries to reduce entropy by finding the splits that result in more **homogeneous** groups.
- **Information gain** is the key metric for selecting the best splits, which is calculated based on the change in entropy before and after the split.




Imagine you have a big box of **colored balls**: some are **red** and some are **blue**. Now, if you reach into the box without looking, you want to guess the color of the ball you pull out.

### High Entropy (Lots of Surprise):
If the box has **half red** balls and **half blue** balls, it’s really hard to guess which color you’ll get. This means there’s a **lot of uncertainty** or **surprise**. You don't know if it will be red or blue because it's equally likely. This situation is called **high entropy** — things are **mixed up** and unpredictable.

### Low Entropy (Little Surprise):
Now, imagine if the box is **almost all red** balls with just **one or two blue** ones. If you reach in, you’re almost certain to pull out a red ball. There's not much surprise, right? You pretty much know what you're going to get. This is called **low entropy** — things are more **organized** and predictable.

### How Entropy Works in a Decision Tree:
When building a decision tree, it's like trying to make the box as predictable as possible. We want the tree to split the balls into smaller groups where each group has **mostly one color**. If we split the balls so one group has almost all red and the other group has almost all blue, the **entropy gets smaller** because now it’s easier to guess the color in each group!

So, **entropy** is just a fancy way of measuring how **mixed up** or **organized** something is. High entropy means lots of mixing and uncertainty, and low entropy means things are more sorted and less surprising!



#### Q11) Using the `predictions` and the `y_test` dataframe calculate the value for each metric using the appropriate function.


In [60]:
#Enter Your Code and Execute

In [61]:
from sklearn.metrics import accuracy_score, jaccard_score, f1_score

Tree_Accuracy_Score = accuracy_score(y_test, predictions)
Tree_JaccardIndex = jaccard_score(y_test, predictions, average='macro')  # Use 'average' based on your classification type
Tree_F1_Score = f1_score(y_test, predictions, average='macro')  # 'macro' for multiclass, 'binary' for binary

# Print the results
print(f"Tree Accuracy Score: {Tree_Accuracy_Score}")
print(f"Tree Jaccard Index: {Tree_JaccardIndex}")
print(f"Tree F1 Score: {Tree_F1_Score}")


Tree Accuracy Score: 0.766412213740458
Tree Jaccard Index: 0.5432397082622094
Tree F1 Score: 0.6843313835366602


- precision (how many of the predicted positives are actual positives) and
- recall (how many of the actual positives were correctly predicted)

### Logistic Regression


### Logistic Regression
**Logistic Regression** is a statistical model used primarily for **binary classification** tasks, meaning it predicts one of two possible outcomes (like Yes/No or 0/1). It is a type of regression model, but unlike **linear regression**, which predicts continuous values, logistic regression predicts probabilities that a given input belongs to a certain class. Based on this probability, the model assigns the input to a class (typically 0 or 1).
#### Key Concepts:
1. **Binary Classification**: Logistic regression is commonly used to classify data into two categories, such as "rain tomorrow or not" or "pass/fail".
2. **Sigmoid Function**: The core of logistic regression is the **sigmoid function** (also known as the logistic function). It maps predicted values to a probability between 0 and 1. The formula for the sigmoid function is:
   \[
   \sigma(z) = \frac{1}{1 + e^{-z}}
   \]
   Where:
   - \(z\) is the linear combination of input features.
   - \(e\) is the base of the natural logarithm.
   If the probability is greater than 0.5, logistic regression will classify the input as 1 (positive class); otherwise, it will classify it as 0 (negative class).
3. **Decision Boundary**: The decision boundary is a threshold that determines the classification. By default, logistic regression uses 0.5 as the decision boundary, meaning inputs with probabilities above 0.5 are classified as 1, and those below 0.5 are classified as 0.
4. **Log-Odds and Linear Equation**: Logistic regression uses the concept of **log-odds** to model the relationship between the input features and the probability of a particular outcome:
   \[
   \log\left(\frac{p}{1 - p}\right) = \beta_0 + \beta_1 x_1 + \beta_2 x_2 + \dots + \beta_n x_n
   \]
   Where:
   - \(p\) is the probability of the outcome (e.g., rain tomorrow).
   - \(\beta_0, \beta_1, \dots, \beta_n\) are the coefficients (parameters) to be learned by the model
### Logistic Regression Explained 
Imagine you're trying to predict something simple, like whether it will rain tomorrow or not. You have some clues, like the temperature, the wind speed, and how cloudy it is today. **Logistic Regression** is like a smart tool that looks at these clues and tries to guess if it’s going to rain.
Here’s how it works:
1. **Guessing with Clues**: Let’s say you know that when it's really cloudy and windy, it usually rains. Logistic Regression looks at clues like this (how cloudy, windy, etc.) and tries to make a smart guess: Will it rain or not?
2. **Probabilities**: Instead of just saying "Yes, it will rain" or "No, it won't", Logistic Regression first gives a **probability**. For example, it might say, “There's a 70% chance it will rain tomorrow.”
3. **Cut-off Point**: Once the model has a probability, it uses a cut-off point. Usually, if the probability is above 50%, it will say "Yes, it will rain", and if it’s below 50%, it will say "No, it won’t rain."
4. **Sigmoid Curve**: To figure out these probabilities, the model uses a special math function called a **sigmoid curve**. It’s a bit like a slide. If you go up the slide, at first you go slow (close to 0% chance), then you speed up in the middle, and then you slow down again (close to 100% chance).
5. **Learning from Data**: Just like how you get better at guessing the weather if you watch it every day, Logistic Regression gets better the more data it has. It looks at past weather data, learns the patterns, and then becomes more accurate.
In short, Logistic Regression is like a clever guesser that uses clues to say, "Yes, it will rain tomorrow" or "No, it won’t," based on what it learned from past clues!

#### Q12) Use the `train_test_split` function to split the `features` and `Y` dataframes with a `test_size` of `0.2` and the `random_state` set to `1`.


In [ ]:
#Enter Your Code and Execute

In [62]:
x_train, x_test, y_train, y_test = train_test_split(features, Y, test_size=0.2, random_state=1, stratify=Y)


#### Q13) Create and train a LogisticRegression model called LR using the training data (`x_train`, `y_train`) with the `solver` parameter set to `liblinear`.


In [ ]:
#Enter Your Code and Execute

In [65]:
LR = LogisticRegression(solver='liblinear', random_state=42)

In [66]:
LR.fit(x_train,y_train)

LogisticRegression(random_state=42, solver='liblinear')


### What is a **Solver**?

A **solver** is like a special calculator that a machine learning model (like Logistic Regression) uses to figure out the best way to make predictions. There are different types of solvers, and each one uses a different method to do this calculation. The solver helps the model find the best way to separate your data into different categories (for example, predicting "yes" or "no").

### What is **liblinear**?

- **liblinear** is one of those special calculators (or solvers) used by the Logistic Regression model.
- It's **good at solving problems when you have small data** or **when you’re only trying to predict between two choices** (like "rain" vs "no rain").
- It works by trying to draw a line (or a boundary) that best separates the two categories.

### Why is **liblinear** useful?

1. **For Smaller Datasets**:  
   Imagine you’re solving a simple puzzle with fewer pieces. You don’t need a big, complicated tool for that, right?  
   - **liblinear** is like a simple tool that works well for small puzzles (small datasets). It can quickly find the best solution without too much effort.

2. **For Binary Classification** (Two Choices):
   - If your problem only has **two possible outcomes**, like "yes" or "no", then **liblinear** works really well because it was designed to handle these two-option problems efficiently.
   
3. **It Can Make Simpler Models**:
   - **liblinear** can help the model decide which features (or pieces of information) are important and which aren’t. It can **ignore unimportant features** by setting their influence to zero. This makes the model simpler and often easier to understand.

### Example:
Imagine you’re building a model to predict whether or not it will rain tomorrow. You have information like temperature, wind speed, and humidity. If:
- You only have a **small amount of data** (maybe just a few hundred days of weather data), and
- You are just predicting **rain (yes/no)**,

Then **liblinear** is like a **quick and reliable tool** to use. It will figure out how much each feature (temperature, wind speed, etc.) influences the prediction, and it will do so efficiently.

### Regularization (Making Predictions Simpler):
- **Regularization** is like a rule that keeps the model from getting too complicated. It helps the model avoid using too many details that don’t really matter.
- **liblinear** supports two types of regularization:
  - **L1 Regularization**: This makes the model simpler by completely removing unimportant features (setting their value to zero).
  - **L2 Regularization**: This makes the model simpler by shrinking the importance of some features without removing them.

---

### Summary:

- **liblinear** is a tool (solver) that works best when:
  - Your dataset is **small**.
  - You are making a prediction with **two possible outcomes** (like yes/no, true/false).
  - You want the model to **ignore** unimportant information (with regularization).
  
This solver is like using a small, quick tool to solve a simple problem instead of using a big, complicated tool that might take more time than needed.



#### Q14) Now, use the `predict` and `predict_proba` methods on the testing data (`x_test`) and save it as 2 arrays `predictions` and `predict_proba`.


In [ ]:
#Enter Your Code and Execute

In [67]:
predictions = LR.predict(x_test)

In [68]:
predictions

array([0., 0., 0., 0., 1., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 1.,
       0., 0., 0., 1., 0., 0., 1., 0., 0., 0., 1., 0., 0., 0., 0., 1., 1.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0.,
       0., 0., 1., 1., 0., 0., 0., 0., 0., 0., 0., 1., 1., 0., 0., 1., 1.,
       0., 1., 1., 1., 0., 1., 1., 0., 0., 1., 0., 0., 1., 1., 0., 1., 0.,
       1., 0., 0., 0., 0., 1., 0., 1., 0., 0., 1., 0., 0., 0., 0., 1., 0.,
       0., 0., 1., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 1., 0., 0., 0., 0., 0., 1., 1., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 1., 0., 0., 1., 1., 1., 0., 0., 0., 0., 1.,
       0., 0., 1., 0., 0., 0., 1., 0., 0., 1., 1., 0., 1., 0., 0., 0., 0.,
       0., 1., 0., 1., 0., 0., 0., 1., 0., 0., 1., 0., 0., 0., 1., 0., 0.,
       0., 0., 0., 0., 1., 0., 0., 0., 1., 0., 1., 0., 0., 1., 0., 0., 0.,
       0., 1., 1., 1., 0.

In [69]:
predict_proba = LR.predict_proba(x_test)

In [72]:
predict_proba[4]

array([0.43217354, 0.56782646])

In [70]:
predict_proba

array([[0.95296047, 0.04703953],
       [0.77225678, 0.22774322],
       [0.98445335, 0.01554665],
       ...,
       [0.94730015, 0.05269985],
       [0.9495956 , 0.0504044 ],
       [0.88554366, 0.11445634]])

Let’s break down what these results mean:

### 1. `predict_proba`

The `predict_proba` function returns the **probability** that each instance in `x_test` belongs to each class. Since logistic regression is a binary classifier (in this case predicting between 0 and 1), the output is a 2-column array:

```python
array([[0.95296047, 0.04703953],
       [0.77225678, 0.22774322],
       [0.98445335, 0.01554665],
       ...
      ])
```

- Each row in this array represents one test sample.
- The two numbers in each row represent the probabilities of the sample belonging to class 0 or class 1:
  - The **first value** is the probability that the sample belongs to class **0**.
  - The **second value** is the probability that the sample belongs to class **1**.

For example:
- `[0.95296047, 0.04703953]` means there's a **95.3% chance** that the sample belongs to class 0, and a **4.7% chance** it belongs to class 1.
- `[0.77225678, 0.22774322]` means a **77.2% chance** for class 0 and a **22.8% chance** for class 1.

### 2. `predict`

The `predict` function uses these probabilities to make **final predictions** for each sample by selecting the class with the **highest probability**. In this case, the model predicts the class as 0 or 1 based on the highest probability value from `predict_proba`.

For example:
- If `predict_proba` gives `[0.95296047, 0.04703953]`, the model will predict class **0** (because 0.9529 is greater than 0.0470).
- If it gives `[0.77225678, 0.22774322]`, it will also predict class **0** (because 0.7722 is greater than 0.2277).

So, the `predict` output:
```python
array([0., 0., 0., 0., 1., 0., 0., 1., 0., ... ])
```
represents the model’s predicted class (0 or 1) for each test sample based on the probabilities.

### Summary:
- **`predict_proba`** gives the **probabilities** for each class (class 0 or 1).
- **`predict`** gives the **final predicted class** (0 or 1) based on the higher probability value from `predict_proba`.



#### Q15) Using the `predictions`, `predict_proba` and the `y_test` dataframe calculate the value for each metric using the appropriate function.


In [ ]:
#Enter Your Code and Execute

In [73]:
LR_Accuracy_Score = accuracy_score(y_test, predictions)
LR_JaccardIndex = jaccard_score(y_test, predictions)
LR_F1_Score = f1_score(y_test, predictions)
LR_Log_Loss = log_loss(y_test, predict_proba)

print(f"LR Accuracy Score: {LR_Accuracy_Score}")
print(f"LR Jaccard Index: {LR_JaccardIndex}")
print(f"LR F1 Score: {LR_F1_Score}")
print(f"LR Log Loss: {LR_Log_Loss}")

LR Accuracy Score: 0.8351145038167939
LR Jaccard Index: 0.4626865671641791
LR F1 Score: 0.6326530612244898
LR Log Loss: 0.37532388067356415


**Log Loss**, also known as **Logarithmic Loss** or **Cross-Entropy Loss**, is a metric used to evaluate the performance of classification models, particularly those that output probabilities. It measures the performance of a model whose predictions are probabilities rather than class labels.

### What Does Log Loss Measure?

Log Loss quantifies how well the predicted probabilities match the true labels. Unlike simple accuracy, which just counts how many predictions are correct, Log Loss takes into account how confident the predictions are. 

- **Lower Log Loss** values indicate that the predicted probabilities are closer to the actual labels.
- **Higher Log Loss** values indicate that the predicted probabilities are further from the actual labels.


### Example

For a binary classification problem:
- If the true label is 1 and the model predicts a probability of 0.9, the Log Loss for that prediction would be low because the predicted probability is close to the actual label.
- If the true label is 1 but the model predicts a probability of 0.1, the Log Loss would be high because the prediction is far from the actual label.

### Why Use Log Loss?

- **Probabilistic Interpretation**: Log Loss provides a more nuanced evaluation of model performance, especially when dealing with probabilities.
- **Penalizes Incorrect Confidences**: It penalizes predictions that are both incorrect and made with high confidence, which helps to avoid overconfidence in the model.

### Summary

Log Loss helps to evaluate how well a classification model’s predicted probabilities match the actual outcomes. It’s particularly useful for models that output probabilities and is often used in competitions and applications where understanding model uncertainty is important.

### SVM



### Here's How It Works:

1. **Drawing the Line**: Think of a Support Vector Machine (SVM) as a smart helper that helps you draw the best line between these two groups of dots. The line should go right in the middle, as far away from the dots of each color as possible.

2. **Best Line**: Imagine that the line can be pushed around. You want to move it to a place where it's farthest from the nearest dots of each color. This way, if a new dot comes in, you’ll be more confident about which side of the line it should go.

3. **Special Dots**: Some dots are very close to the line, and these are called “support vectors.” They are important because they help decide where the line should go.

4. **Tricky Shapes**: Sometimes, dots are mixed up and can’t be separated by just a straight line. In those cases, the SVM can use a trick to turn the paper into a 3D surface (like bending the paper) so you can draw a line that separates the dots better.

### Why Use SVMs?

- **Good at Separating Dots**: SVMs are really good at drawing a clear line between different groups of dots, even if there are lots of dots or the dots are in tricky shapes.
- **Handles Complicated Situations**: If the dots are all mixed up and you can’t draw a straight line, SVMs can use clever tricks to separate them.

So, an SVM is like a super-smart tool that helps you draw the best line to divide groups of things, even if it’s not always straightforward!

You want to draw the line in such a way that it separates the two colors as clearly as possible.

### Here's How It Works:

1. **Drawing the Line**: Think of a Support Vector Machine (SVM) as a smart helper that helps you draw the best line between these two groups of dots. The line should go right in the middle, as far away from the dots of each color as possible.

2. **Best Line**: Imagine that the line can be pushed around. You want to move it to a place where it's farthest from the nearest dots of each color. This way, if a new dot comes in, you’ll be more confident about which side of the line it should go.

3. **Special Dots**: Some dots are very close to the line, and these are called “support vectors.” They are important because they help decide where the line should go.

4. **Tricky Shapes**: Sometimes, dots are mixed up and can’t be separated by just a straight line. In those cases, the SVM can use a trick to turn the paper into a 3D surface (like bending the paper) so you can draw a line that separates the dots better.

### Why Use SVMs?

- **Good at Separating Dots**: SVMs are really good at drawing a clear line between different groups of dots, even if there are lots of dots or the dots are in tricky shapes.
- **Handles Complicated Situations**: If the dots are all mixed up and you can’t draw a straight line, SVMs can use clever tricks to separate them.

So, an SVM is like a super-smart tool that helps you draw the best line to divide groups of things, even if it’s not always straightforward!

Let’s connect Support Vector Machines (SVMs) to predicting rain, using a simple example.

### Scenario

Imagine you want to predict whether it will rain tomorrow based on several features like temperature, humidity, and wind speed. You have historical weather data where each day is labeled as "Rain" or "No Rain."

### Using SVM for Rain Prediction

1. **Features and Labels**:
   - **Features**: These are the pieces of information you use to make predictions, such as temperature, humidity, and wind speed.
   - **Labels**: These are the outcomes you’re trying to predict, like whether it rained or not.

2. **Training the SVM**:
   - **Draw the Line (or Hyperplane)**: SVM will try to find the best way to separate the days with rain from the days without rain. In our case, it will look for a line (or in higher dimensions, a plane or hyperplane) that best divides the days into two groups: "Rain" and "No Rain."
   - **Maximize the Margin**: SVM will place this line in a way that it is as far as possible from the closest days (data points) in each group. This ensures that the separation is as clear as possible.

3. **Kernel Trick (if needed)**:
   - Sometimes, the relationship between features and labels isn't a simple straight line. For example, maybe rain depends on a combination of temperature and humidity in a non-linear way.
   - **Transforming the Data**: The SVM can use a technique called the kernel trick to handle more complex relationships. This might involve turning your features into a higher-dimensional space where a clear separation becomes possible.

4. **Making Predictions**:
   - Once the SVM model is trained, you can use it to predict whether it will rain tomorrow based on the new weather data you have.
   - **Classify New Data**: When you get new data (like tomorrow’s temperature, humidity, and wind speed), the SVM will use the line or hyperplane it learned during training to decide whether it predicts "Rain" or "No Rain."

### Why Use SVM for Rain Prediction?

- **Effective in Complex Situations**: SVMs are good at finding clear boundaries between classes even when the data is complex.
- **Handles High Dimensions**: SVMs work well with many features, which is useful if you have a lot of different types of weather data.

### Summary

Using SVM for rain prediction is like drawing the best possible line or boundary to separate days with rain from days without rain. By making sure this boundary is as clear as possible, the SVM helps you predict whether it will rain tomorrow based on the current weather conditions.

#### Q16) Create and train a SVM model called SVM using the training data (`x_train`, `y_train`).


In [74]:
#Enter Your Code and 
from sklearn.svm import SVC


In [85]:
SVM = SVC(kernel='rbf', random_state=42)

In [86]:
SVM.fit(x_train,y_train)

SVC(random_state=42)

#### Q17) Now use the `predict` method on the testing data (`x_test`) and save it to the array `predictions`.


In [87]:
#Enter Your Code and Execute

In [88]:
predictions = SVM.predict(x_test)

In [89]:
predictions

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0.

#### Q18) Using the `predictions` and the `y_test` dataframe calculate the value for each metric using the appropriate function.


In [90]:
# Calculate the metrics
SVM_Accuracy_Score = accuracy_score(y_test, predictions)
SVM_JaccardIndex = jaccard_score(y_test, predictions)
SVM_F1_Score = f1_score(y_test, predictions)

# Print the results
print(f"SVM Accuracy Score: {SVM_Accuracy_Score}")
print(f"SVM Jaccard Index: {SVM_JaccardIndex}")
print(f"SVM F1 Score: {SVM_F1_Score}")

SVM Accuracy Score: 0.7404580152671756
SVM Jaccard Index: 0.0
SVM F1 Score: 0.0


### Report


#### Q19) Show the Accuracy,Jaccard Index,F1-Score and LogLoss in a tabular format using data frame for all of the above models.

\*LogLoss is only for Logistic Regression Model


In [97]:
results_df = pd.DataFrame({
    'Model': ['Linear Regression', 'KNN', 'Decision Tree', 'Logistic Regression', 'SVM'],
    'MAE': [LinearRegression_MAE, None, None, None, None],
    'MSE': [LinearRegression_MSE, None, None, None, None], 
    'R2': [LinearRegression_R2, None, None, None, None],
    'Accuracy Score': [None, KNN_Accuracy_Score, Tree_Accuracy_Score, LR_Accuracy_Score, SVM_Accuracy_Score],
    'Jaccard Index': [None, KNN_JaccardIndex, Tree_JaccardIndex, LR_JaccardIndex, SVM_JaccardIndex],
    'F1 Score': [None, KNN_F1_Score, Tree_F1_Score, LR_F1_Score, SVM_F1_Score],
    'Log Loss': [None, None, None, LR_Log_Loss, None]  # Log Loss is only applicable for Logistic Regression
})

In [98]:
Report = results_df

In [99]:
Report

Model       MAE       MSE        R2  Accuracy Score  \
0    Linear Regression  0.261357  0.127618  0.335944             NaN   
1                  KNN       NaN       NaN       NaN        0.815267   
2        Decision Tree       NaN       NaN       NaN        0.766412   
3  Logistic Regression       NaN       NaN       NaN        0.835115   
4                  SVM       NaN       NaN       NaN        0.740458   

   Jaccard Index  F1 Score  Log Loss  
0            NaN       NaN       NaN  
1       0.366492  0.536398       NaN  
2       0.543240  0.684331       NaN  
3       0.462687  0.632653  0.375324  
4       0.000000  0.000000       NaN

<h2 id="Section_5">  How to submit </h2>

<p>Once you complete your notebook you will have to share it. You can download the notebook by navigating to "File" and clicking on "Download" button.

<p>This will save the (.ipynb) file on your computer. Once saved, you can upload this file in the "My Submission" tab, of the "Peer-graded Assignment" section.  


<h2>About the Authors:</h2> 

<a href="https://www.linkedin.com/in/joseph-s-50398b136/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkML0101ENSkillsNetwork20718538-2022-01-01">Joseph Santarcangelo</a> has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.

### Other Contributors

[Svitlana Kramar](https://www.linkedin.com/in/svitlana-kramar/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMML0232ENSkillsNetwork30654641-2022-01-01)


## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>

<!--
## Change Log

| Date (YYYY-MM-DD) | Version | Changed By    | Change Description          |
| ----------------- | ------- | ------------- | --------------------------- |
| 2022-06-22        | 2.0     | Svitlana K.   | Deleted GridSearch and Mock |
--!>